lib/orangelight/browse_lists/call_number_csv.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+module BrowseLists
+  class SolrResponseError < StandardError; end
+  class CallNumberCSV
+    attr_reader :output_root, :facet_request, :conn, :rows
+    attr_accessor :multi_cn_lookup
+    def initialize(facet_request, conn, output_root: Pathname.new("/tmp"), rows: 5_000)
+      @output_root = output_root
+      @facet_request = facet_request
+      @conn = conn
+      @rows = rows
+      @multi_cn_lookup = {}
+    end
+
+    def write
+      write_multiple_call_numbers
+      write_remaining_call_numbers
+    end
+
+    def filename
+      output_root.join("#{facet_field}.csv")
+    end
+
+    private
+
+      def core_url
+        Blacklight.default_index.connection.uri.to_s.gsub(%r{^.*\/solr}, '/solr')
+      end
+
+      def facet_field
+        "call_number_browse_s"
+      end
+
+      # Collect all the call numbers with facet count > 2
+      # Write each as a row in the csv and add it to the lookup table
+      def write_multiple_call_numbers
+        # This request takes a minute or so
+        resp = conn.get "#{facet_request}#{facet_field}&facet.mincount=2"
+        req = JSON.parse(resp.body)
+        CSV.open(filename, 'wb') do |csv|
+          req['facet_counts']['facet_fields'][facet_field].each_slice(2) do |mcn, record_count|
+            sort_cn = StringFunctions.cn_normalize(mcn)
+            multi_cn_lookup[sort_cn] = record_count
+            csv << [sort_cn, mcn, 'ltr', '', "#{record_count} titles with this call number", '', '', "?f[#{facet_field}][]=#{CGI.escape(mcn)}", '', 'Multiple locations']
+          end
+        end
+      end
+
+      # Append the rest of the call numbers to the file
+      def write_remaining_call_numbers
+        cursor_mark = '*'
+        CSV.open(filename, 'ab') do |csv|
+          loop do
```

Comment:
Suggestion: It seems like this loop logic along with solr_cursor_page_body could be extracted to an enumerator, so this logic could just become
```ruby
CSV.open(filename, 'ab') do |csv|
  call_number_records.each do |record|
    csv << record.row
  end
end
```

rather than having it all be procedural.